File: solutions/day23.js
const fs = require("fs");

function pair_equal(x, y){
    return (x[0] == y[0]) && (x[1] == y[1]);
}

function parse(lines){
    return [...lines.map((x) => x.split(""))];
}

function get_neighbors(graph){
    let xmin = ymin =0;
    let xmax = graph[0].length - 1;
    let ymax = graph.length -1;
    //let result = [...Array( ymax).map(() => Array(xmax))]
    let result = []
    let map = {};
    // Clockwise order
    let iterations = {"." : [[0, -1], [1, 0], [0, 1], [-1, 0]], "#" : [], "<" : [[-1, 0], [0, -1], [1, 0], [0, 1]],
        "^" : [[0, -1], [1, 0], [0, 1], [-1, 0]],
        ">" : [[1, 0], [0, 1], [-1, 0], [0, -1]],
        "v" : [[0, 1], [-1, 0], [0, -1], [1, 0]]}

    for(let row = ymin; row <= ymax; row++){
        result.push([]);
        for(let col =xmin; col <= xmax; col ++){
            result[row].push([]);
            let current = [row, col];
            let char = graph[row][col];
            let shifts = iterations[char];
            if (graph[row][col] != "#"){
            //result[row][col] = [];
            for(dir of shifts){
                let current = [row + dir[0], col + dir[1]];
                if (((current[1] >= xmin) && current[1] <= xmax) && ((current[0] >= ymin) && current[0] <= ymax) && graph[current[0]][current[1]] != "#"){
                    result[row][col].push(current);
                }
            }

            }
        }
    }
    return result;
}

function find_start(lines){ 
    let endpoint = lines.length -1;
    for(let i =0; i< lines[0].length; i++){ 
        if (lines[0][i] == "."){ 
            let start = [0, i];
            for(let j =0; j< lines[endpoint].length; j++){ 
                if (lines[endpoint][j] == "."){ 
                    return [start, [endpoint, j]]
                }
        }
    }
    }
}

function dijkstra(start, goal, neighbors){ 
    let queue = [[new Set([]), JSON.stringify(start)]];
    let target = JSON.stringify(goal)
    let best = -Infinity;

    while (queue.length){ 
        let current = queue.pop();
        let previous = current[0]
        
        console.log(current[1])
        if (current[1] == target) { 
            best = Math.max(previous.size, best);
            console.log(best)
            continue
        }
        let current_coord = JSON.parse(current[1]);
        for(neighbor of neighbors[current_coord[0]][current_coord[1]]){
            let string = JSON.stringify(neighbor); 
            //console.log(string)
            if (!(previous.has(string))){ 
                let copy = new Set(previous);
                copy.add(current[1])
                queue.push([copy, string]);
            }
    }
}
    return best ; 
}



const raw_input = fs.readFileSync('inputs/day23.txt', 'utf-8').toString().replace(/\n+$/, "").split("\n");
const graph = parse(raw_input);
let iterations = {"." : [[0, -1], [1, 0], [0, 1], [-1, 0]], "#" : [], "<" : [[-1, 0], [0, -1], [1, 0], [0, 1]],
        "^" : [[0, -1], [1, 0], [0, 1], [-1, 0]],
        ">" : [[1, 0], [0, 1], [-1, 0], [0, -1]],
        "v" : [[0, 1], [-1, 0], [0, -1], [1, 0]]}
const endpoints = find_start(graph);
const neighbors = get_neighbors(graph);
const part2 = dijkstra(endpoints[0], endpoints[1], neighbors);
console.log(part2)
// Part 1 restriction
for (char of ["<", "^", ">", "v"] ){ 
    iterations[char] = [ iterations[char] ][0]
}
const new_neighbors = get_neighbors(graph);
const part1 = dijkstra(endpoints[0], endpoints[1], new_neighbors);
console.log(part1)

File: solutions/day23.py
from collections import defaultdict
from collections import deque
from functools import cache
from math import inf

from utils.utils import split_lines


def print_path(path, graph):
    return "\n".join(
        "".join(
            graph[complex(x, y)] if complex(x, y) not in path else "O"
            for x in range(23)
        )
        for y in range(23)
    )


def get_path(graph, start):
    current = start
    queue = []
    inside = True

    while inside:
        queue.append(current)
        for dir in {-1, 1, 1j, -1j}:
            new = dir + current
            try:
                char = graph[int(new.imag)][int(new.real)]
            except:
                continue
            if char == "O" and new not in queue:
                current += dir
                break
        else:
            inside = False

    return queue


def mark_path(lines):
    return {
        complex(x, y)
        for y, line in enumerate(lines)
        for x, char in enumerate(line)
        if char in {"S", "O"}
    }


def find_start(lines, reverse=False):
    loop = reversed(list(enumerate(lines))) if reverse else enumerate(lines)
    for y, line in loop:
        for x, char in enumerate(line):
            if char == ".":
                return complex(x, y)


def parse(lines):
    return {
        complex(x, y): char
        for y, line in enumerate(lines)
        for x, char in enumerate(line)
    }


def make_neighbors(graph, directions):
    @cache
    def result(coord):
        result = set()
        offsets = directions[graph[coord]]
        for offset in offsets:
            new = coord + offset
            if new in graph and graph[new] != "#":
                result.add(new)
        return frozenset(result)

    return result


def node_paths(start, neighbors, targets):
    queue = deque([[start]])

    while queue:
        current = queue.pop()
        current_coord = current[-1]

        for neighbor in neighbors(current_coord):
            if neighbor not in current:
                # New path found to node, so end here
                if neighbor in targets and neighbor != current_coord:
                    # Longer path found
                    yield start, neighbor, len(current)
                else:
                    queue.append(current + [neighbor])


def reduce_graph(graph, start, goal, neighbors):
    targets = {c for c  in graph.keys() if len(neighbors(c)) > 2} | {start, goal}
    result = defaultdict(dict)
    for target in targets:
        gen = node_paths(target, neighbors, targets)
        # Exhaust results from generator
        while True:
            try:
                this = next(gen)
                if this is not None:
                    start, dest, length = this
                    result[start][dest] = max(result[start].get(dest, -inf), length)
                    result[dest][start] = max(result[dest].get(start, -inf), length)
            except StopIteration:
                break
    return result


def solve(graph, start, goal):
    queue = deque([(set(), start,  0)])
    result = -inf

    while queue:
        current, current_coord, current_dist = queue.pop()
        if current_coord == goal:
            result = max(result, current_dist)
        else:
            neighbors = graph[current_coord]
            for neighbor, dist in neighbors.items():
                if neighbor not in current:
                    new_dist = current_dist + dist
                    queue.append((current | {current_coord}, neighbor, new_dist))
    return result


def dijkstra(graph, start, goal, threshold, neighbors):
    queue = deque([(set(), start)])
    part1 = threshold

    while queue:
        previous, current_coord = queue.pop()
        if current_coord == goal:
            part1 = max(len(previous), part1)
            continue

        for neighbor in neighbors(current_coord):
            if neighbor not in previous:
                queue.append((previous | { current_coord }, neighbor))

    return part1 


raw_input = split_lines("inputs/day23.txt")
directions = {
    ".": frozenset({-1, 1, -1j, 1j}),
    "^": frozenset({-1j}),
    ">": frozenset({1}),
    "v": frozenset({1j}),
    "<": frozenset({-1}),
    "#": frozenset(),
}
start = find_start(raw_input)
goal = find_start(raw_input, reverse=True)
graph = parse(raw_input)
neighbors = make_neighbors(graph, directions)
part1 = dijkstra(graph, start, goal, -inf, neighbors)
print(part1)


directions[">"] = directions["v"] = directions["<"] = directions["^"] = directions["."]
neighbors = make_neighbors(graph, directions)
reduced = reduce_graph(graph, start, goal, neighbors)
part2 = solve(reduced, start, goal)
print(part2)
